### PR TITLE
Make PasswordHasher use Argon2id

### LINKED ISSUE #33: Argon2id should be the default
Currently `Argon2i` is used as the default algorithm. There are [two published attacks](https://en.wikipedia.org/wiki/Argon2#Cryptanalysis) against it. The [recommended](https://tools.ietf.org/html/draft-irtf-cfrg-argon2-03) and primary variant of `Argon2` is `Argon2id` (also default in libsodium).

### PR DESCRIPTION
**Pull request by hynek**
Also introduce hash type agility to stay backward compatible.

C.f. https://tools.ietf.org/html/draft-irtf-cfrg-argon2-03#section-4

Fixes #33

### CODE REVIEW

File: docs/api.rst [hunk 1]
```diff
@@ -13,7 +13,7 @@ Unless you have any special needs, all you need to know is:
   >>> ph = PasswordHasher()
   >>> hash = ph.hash("s3kr3tp4ssw0rd")
   >>> hash  # doctest: +SKIP
-  '$argon2i$v=19$m=512,t=2,p=2$5VtWOO3cGWYQHEMaYGbsfQ$AcmqasQgW/wI6wAHAMk4aQ'
+  '$argon2id$v=19$m=512,t=2,p=2$Wd3CJItnL93pSLG/Mvel2g$DnsqcrS7+Enwu8EdJrUX2Q'
```
Comment (lvh) [inline docs/api.rst]: sanity check: I'm assuming the contract only ever was "pass me a string back that I once gave you to begin with"?

File: docs/parameters.rst [hunk 2]
```diff
@@ -4,24 +4,32 @@ Choosing Parameters
...
-The authors of Argon2 specified a method in their `paper <https://github.com/P-H-C/phc-winner-argon2/blob/master/argon2-specs.pdf>`_ but it should be noted that they also  mention that no value for ``time_cost`` or ``memory_cost`` is actually insecure (cf. section 6.4).
+The authors of Argon2 specified a method in their `paper <https://github.com/P-H-C/phc-winner-argon2/blob/master/argon2-specs.pdf>`_, however some parts of it have been revised in the `RFC draft`_ for Argon2 that is currently being written.
 
-#. Choose whether you want Argon2i or Argon2d (``type``).
-   If you don't know what that means, choose Argon2i (:attr:`argon2.Type.I`).
-#. Figure out how many threads can be used on each call to Argon2 (``parallelism``).
+The current recommended best practice is as follow:
+
+#. Choose whether you want Argon2i, Argon2d, or Argon2id (``type``).
+   If you don't know what that means, choose Argon2id (:attr:`argon2.Type.ID`).
```
Comment (lvh) [inline docs/parameters.rst]: :100:

File: src/argon2/_password_hasher.py [hunk 3]
```diff
@@ -26,10 +27,11 @@ class PasswordHasher(object):
     r"""
     High level class to hash passwords with sensible defaults.
 
-    Uses *always* Argon2\ **i** and a random salt_.
+    Uses *always* Argon2\ **id** and a random salt_ for hashing, but it can
```
Comment (lvh) [inline src/argon2/_password_hasher.py]: Suggestion: always uses, not uses always

File: src/argon2/_password_hasher.py [hunk 4]
```diff
@@ -122,9 +139,16 @@ def verify(self, hash, password):
         .. versionchanged:: 16.1.0
             Raise :exc:`~argon2.exceptions.VerifyMismatchError` on mismatches
             instead of its more generic superclass.
+        .. versionadded:: 18.2.0 Hash type agility.
         """
+        hash = _ensure_bytes(hash, "ascii")
+        try:
+            hash_type = self._header_to_type[hash[:9]]
+        except (IndexError, KeyError):
```
Comment (lvh) [inline src/argon2/_password_hasher.py]: Suggestion: LookupError

File: docs/parameters.rst [hunk 5]
```diff
@@ -4,24 +4,32 @@ Choosing Parameters
...
 #. Figure out how much memory each call can afford (``memory_cost``).
-#. Choose a salt length.
-   16 Bytes are fine.
-#. Choose a hash length (``hash_len``).
-   16 Bytes are fine.
+   The RFC recommends 4 GB for backend authentication and 1 GB for frontend authentication.
+#. Select the salt length.
+   16 bytes is sufficient for all applications, but can be reduced to 8 bytes in the case of space constraints.
+#. Choose a hash length (``hash_len``, called "tag length" in the documentation).
+   16 bytes is sufficient for most applications, including key derivation.
```
Comment (lvh) [inline docs/parameters.rst]: I usually tell people to derive 256 bit symmetric keys, but you don't have to edit that. 16 is fine for password verification, definitely. Since the output of the derivation function has a header in it you would probably still wring it through HKDF or something.

File: docs/api.rst (hunk 1 above)
Comment (hynek) [reply docs/api.rst]: There's not contract here really because the salt is random. So I basically created one by hand and put it there.

File: docs/parameters.rst (hunk 5 above)
Comment (hynek) [reply docs/parameters.rst]: I’ll rephrase it to make it about password verification.

File: src/argon2/_password_hasher.py (hunk 3 above)
Comment (hynek) [reply src/argon2/_password_hasher.py]: Gah I *always* get this one wrong.

**Comment by hynek** [comment]
> So what happens when an attacker does get to mess with that record?

I’ve added a blurb reinforcing that argon2_cffi does not expect malicious hashes.

> Additionally, you may want to make the types parameterizable and expose the fact that the password is not complaint with the current standard (be it type or either kind of cost) via an API, documenting the way people could upgrade their existing password stores that way

I’ve opened #35 and #36 for that if I understand you correctly.

> As an aside, I'm not happy that somehow we had a PHC to finally get this right and apparently the answer is "choose between these 3 subtly different things based on specialized knowledge you're not supposed to have".

Ironically, it seems like we’ve just arrived there with Argon2id. :)

***

I made some updates (including stressing some more, that (ye)script is fine. It would be great if you could give it one last glance!

**Comment by hynek** [comment]
The last remaining question for me is the recommended hash verification time. Some people say 0.5ms others 500ms which is…quite the spread. :| I don't run anything at massive scale, so I'm not sure what to recommend?

**Comment by lvh** [comment]
Re: #35, #36: yep, that's a start, but not the entirety of it. One part is "given a token I am willing to accept and a password that hashes to it, should I re-hash the password to these new parameters?". That's #36 and #35 is a necessary prereq. The other question is: "given this token, should I be willing to attempt to validate a password against it at all"? I.e. what happens when the token tells me to allocate eleventy gajillion jiggabytes? That doesn't seem captured in either ticket.

Re: 5ms vs 500ms,more is better, but UX is the limiting factor. Some guidance:

* No setting will save you from an awful password. At 500ms, the top 10k passwords still only take me less than two hours to try if I have no concurrency at all.
* 500ms is pretty high. .5ms seems low. I get what you're saying, don't have ranges. I'd suggest 50ms as a default, but document that people have suggested anywhere from .5ms (the blog post) to 500ms (the RFC), and the real answer is "more is better and UX is the limiting factor", and 50ms is a good trade-off for most people. (Similarly I think the memory default is pretty high, but I'm OK with secure defaults :))

**Comment by lvh** [comment]
Some more concrete advice: libsodium has this nice thing where they have presets:

**Comment by hynek** [comment]
I’ve added a clarification – thanks for you help! <3
